File: 矩阵理论与方法/课件/第13次课12.13/代码实现参考/代码实现参考/第一题/test1.py
# -*- coding: utf-8 -*-
'''
求解Ax=b
'''

from __future__ import division
import time
import numpy as np  
from scipy import linalg as la 

N = 3000
b = np.random.random([N,1])

A = np.zeros([N,N])
a0 = np.random.random([1,N])
a1 = np.hstack((a0,a0))

for k in range(0,N):
    A[k,:] = a1[0, N-k:2*N-k]

time0 = time.process_time()


#高斯消去法
x0 = np.linalg.solve(A, b)
time1 = time.process_time()


#scipy版本plu分解
P, L, U = la.lu(A) #A=PLU
z = np.linalg.solve(P, b) #Pz = b 求z
y = np.linalg.solve(L, z) #Ly=z 求y
x1 = np.linalg.solve(U, y) #Ux=y 求x
'''
#补充, 取p为P的逆，就改成了通常所用的列主元LU分解  pA=LU
p = np.linalg.inv(P) #取p为P的逆, pA=LU
y = np.linalg.solve(L, p.dot(b)) #Ly=pb 求y
x1 = np.linalg.solve(U, y) #Ux=y 求x
#'''
time2 = time.process_time()


#循环矩阵可以用fft
Lambda = np.fft.fftn(A[:,0])
inv_lambda = np.diag(1/Lambda)
x2 = np.fft.ifftn(np.dot(inv_lambda, np.fft.fftn(b)))
time3 = time.process_time()


time_gauss = time1 - time0
time_plu = time2 - time1
time_fft = time3 - time2

print('time_gauss',time_gauss)
print('time_plu',time_plu)
print('time_fft',time_fft)
print('error',np.linalg.norm(x0-x1) + np.linalg.norm(x0-x2))

File: 矩阵理论与方法/课件/第13次课12.13/代码实现参考/代码实现参考/第一题/test1.m
clear

N=3000;
b=rand(N,1);             %随机生成向量b

A=zeros(N);
a0=rand(N,1);
for k=0:N-1
   A(:,k+1)=circshift(a0,k);        %随机生成循环矩阵A
end

% 求解Ax=b

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
time0=cputime;

x_inv=inv(A)*b;             %直接求逆计算

time1=cputime;
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

x_gauss=A\b;             %高斯消去法计算

time2=cputime;
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

[L,U,P]=lu(A);
x_PLU=U\(L\(P*b));             %PLU分解计算，列主元LU分解，满足PA=LU

time3=cputime;
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

x_fft=ifft(diag(1./fft(A(:,1)))*fft(b));         %循环矩阵可以利用快速傅里叶变换计算

time4=cputime;
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

error=norm(x_inv-x_fft)+norm(x_inv-x_gauss)+norm(x_inv-x_PLU);         %检验误差
time_inv=time1-time0;
time_gauss=time2-time1;
time_plu=time3-time2;
time_fft=time4-time3;

% fprintf('矩阵A条件数为%f\n', cond(A));
fprintf('直接求逆计算时间为%f\n', time_inv);
fprintf('高斯消去法计算时间为%f\n', time_gauss);
fprintf('PLU分解计算时间为%f\n', time_plu);
fprintf('FFT计算时间为%f\n', time_fft);

error
